tests/gflownet/envs/common.py

```diff
@@ -12,47 +13,484 @@
+            )
+            assert all([action == self.env.eos for action in actions])
+
+    def test__get_logprobs__backward__returns_zero_if_done(self, n_repeat=1):
+        N = 5
+
+        if _get_current_method_name() in self.repeats:
+            n_repeat = self.repeats[_get_current_method_name()]
+
+        for _ in range(n_repeat):
+            states = _get_terminating_states(self.env, N)
+            if states is None:
+                warnings.warn("Skipping test because states are None.")
+                return
+            # Set states, done and get masks
+            masks = []
+            for state in states:
+                self.env.set_state(state, done=True)
+                masks.append(self.env.get_mask_invalid_actions_backward())
+            # EOS actions
+            actions_eos = torch.tile(
+                tfloat(self.env.eos, float_type=self.env.float, device=self.env.device),
+                (len(states), 1),
+            )
+            # Build random policy outputs and tensor masks
+            policy_outputs = torch.tile(self.env.random_policy_output, (len(states), 1))
+            # Add noise to policy outputs
+            policy_outputs += torch.randn(policy_outputs.shape)
+            masks = tbool(masks, device=self.env.device)
+            logprobs = self.env.get_logprobs(
+                policy_outputs, actions_eos, masks, states, is_backward=True
+            )
+            assert torch.all(logprobs == 0.0)
+
+    def test__forward_actions_have_nonzero_backward_prob(self, n_repeat=1):
+        if _get_current_method_name() in self.repeats:
+            n_repeat = self.repeats[_get_current_method_name()]
+
+        for _ in range(n_repeat):
+            self.env.reset()
+            policy_random = torch.unsqueeze(self.env.random_policy_output, 0)
+            while not self.env.done:
+                state_next, action, valid = self.env.step_random(backward=False)
+                if not valid:
+                    continue
+
+                # Get backward logprobs
+                mask_bw = self.env.get_mask_invalid_actions_backward()
+                masks = torch.unsqueeze(tbool(mask_bw, device=self.env.device), 0)
+                actions_torch = torch.unsqueeze(
+                    tfloat(action, float_type=self.env.float, device=self.env.device), 0
+                )
+                policy_outputs = policy_random.clone().detach()
+                logprobs_bw = self.env.get_logprobs(
+                    policy_outputs=policy_outputs,
+                    actions=actions_torch,
+                    mask=masks,
+                    states_from=[self.env.state],
+                    is_backward=True,
+                )
+                assert torch.isfinite(logprobs_bw)
+                assert logprobs_bw > -1e6
+                state_prev = copy(state_next)  # TODO: We never use this. Remove?
+
+    def test__backward_actions_have_nonzero_forward_prob(self, n_repeat=1):
+        N = 1000
```

Comment:
How much work would it be to make `N` (or `n_states`) an argument like `n_repeat`, so that it could be modified in a dictionary passed from each environment's specific class? It seems that it would be equivalent to `n_repeat`, right? Or am I missing something?

Reply:
done!